### ISSUE TITLE
support specifying extra attributes for `a` tags

### ISSUE THREAD

**Issue by invakid404** [toboshii/hajimari#48]
**Is your feature request related to a problem? Please describe.**
If I try to go to my self-hosted qBittorrent instance through Hajimari, I get hit with an "unauthorized" due to the CSRF protection of qBittorrent. More specifically, because of the requirement that the Referer and Target headers must be the same:

```
(W) 2022-09-28T09:47:04 - WebUI: Referer header & Target origin mismatch! Source IP: '::ffff:10.42.0.16'. Referer header: 'https://lab.invak.id/'. Target origin: 'qbit.lab.invak.id'
```

Note: `lab.invak.id` is Hajimari, and `qbit.lab.invak.id` is my qBittorrent instance.

**Describe the solution you'd like**
I'd like to have a way of specifying `rel="noreferrer"` on the link. Speaking more generally, I think it makes sense to have a way to add extra attributes to the `a` tags. For example, it could be useful for opening links in a new tab (`target="_blank"`) as well.

**Describe alternatives you've considered**
For my specific case, an alternative would be to disable the CSRF protection of qBittorrent. Still, I'm not really a fan of the idea of deliberately reducing security settings in order to be able to have my qBittorrent instance bookmarked. And I do think that my feature request would be useful in the general case as well.

**Additional context**
FWIW, here are my full Hajimari manifests: [link to my home cluster repo](https://github.com/invakid404/home-cluster/tree/main/apps/hajimari)

**Comment by ullbergm** [comment]
@toboshii
I was working on a PR to add the ability to change the link target for the `a` tags with a boolean ("" or "_blank" depending on the value), but I can pivot and add an annotation to add arbitrary html code to the app.url links and cover this issue if you prefer that solution.

If so, any preferences for the naming of the annotation? 

Other option is to make both these use-cases boolean toggles or strings rather than allow arbitrary html code.

**Comment by toboshii** [comment]
From my testing adding `rel="noopener noreferrer"` makes no difference on fixing the qBittorrent CSRF issue, unless you also include `target="_blank"` (in Firefox at least).

I'm a little against end users being able to inject random attributes into html unless there's a really good reason so I'm leaning more towards a simple boolean that toggles `target="_blank" rel="noopener noreferrer"`. According to several of the links and comments [here](https://stackoverflow.com/questions/50709625/link-with-target-blank-and-rel-noopener-noreferrer-still-vulnerable), most if not all modern browsers automatically include `noopener` on any blank targets, and `noreferrer` includes the functionality of `noopener`, but no harm in including  both just to be sure.

That is unless someone has some really good use cases for other anchor attributes, or a valid use case where always including noreferrer on a blank target is an issue.

As far as configuration and naming I'd like to see an annotation (perhaps `hajimari.io/targetBlank`?) that controls it per app, as well as a global toggle (`linksTargetBlank`?) that affects all app and bookmark links. Open to any suggestions here though.

**Comment by invakid404** [comment]
Odd, `rel="noreferrer"` seems to be enough on my end in both Vivaldi and Firefox:

https://user-images.githubusercontent.com/38114607/193315852-9be6c252-61c5-41ae-b4ae-2179c97f56a6.mp4

Could it be caused by some difference in our configs?

**Comment by ullbergm** [comment]
I agree that random html could be iffy, no telling what could break and it is not a great security practice.

Initially I started out allowing different targets but I saw that the functionality had been deprecated in HTML5, only "" or _blank are valid options going forward.

* Adding  `linksTargetBlank`, defaulting to `false` in order to not change current behavior.
* Apps and Bookmarks get a `targetBlank` attribute that will override the global default.
* When set to `true` it adds `target="_blank" rel="noopener noreferrer"` to the app/bookmark `a` tags.

Or alternatively we split the `targetBlank` from the `noreferrer` thing in case someone needs the granularity?

**Comment by invakid404** [comment]
I personally would like a bit more granularity than a single annotation that toggles both `target="_blank"` and `rel="noopener noreferrer"`, I think it makes sense to have two separate annotations at the very least, as I personally only need the latter and I'd prefer if my links still open in the same tab, considering it works fine for me (and for plenty of other people as well I'd assume).

**Comment by ullbergm** [comment]
Is there a drawback to adding `noopener noreferrer` to the links by default?

**Comment by ullbergm** [comment]
I saw you added the noreferrer by default so I created a PR for just the targetBlank option - #49

**Comment by toboshii** [comment]
> Odd, `rel="noreferrer"` seems to be enough on my end in both Vivaldi and Firefox:

Ah, thanks for confirming, may be something to do with me having Authelia in the mix.

> Is there a drawback to adding `noopener noreferrer` to the links by default?

Yea I can't think of a scenario where having these as default breaks something, so lets just have it default and we can add a flag later if someone reports a problem.

**Comment by toboshii** [comment]
Added in #44

**Comment by invakid404** [comment]
@toboshii now that there's a release of the Svelte version and I could test it properly, I can see why you said that `rel="noreferrer"` does nothing: because it actually doesn't. The detail we overlooked is that SvelteKit's client-side router gets in the way. The fix is basically to add `rel="external"` as well, which effectively disables the SvelteKit client-side router for those anchor tags. I've opened a PR with the fix: https://github.com/toboshii/hajimari/pull/50
